pyphi/tpm.py

```diff
@@ -3,16 +3,283 @@
 # tpm.py
 
 """
-Functions for manipulating transition probability matrices.
+Provides the TPM, ExplicitTPM, and ImplicitTPM classes.
 """
 
 from itertools import chain
 
 import numpy as np
 
-from .convert import to_multidimensional
+from . import config, convert, exceptions
+from .utils import all_states, np_hash, np_immutable
 from .constants import OFF, ON
-from .utils import all_states
+
+
+class TPM:
+    """A transition probability matrix."""
+    def __init__(self, tpm, validate=True):
+        self._tpm = self._build_tpm(tpm, validate)
+
+    @property
+    def tpm(self):
+        """Return the underlying `tpm` object."""
+        return self._tpm
+
+    def _build_tpm(self, tpm, validate=True):
+        """Validate the TPM passed by the user and convert to
+        multidimensional form."""
+        raise NotImplementedError
+
+    def conditionally_independent(self):
+        """Validate that this TPM is conditionally independent."""
+        return self._conditionally_independent(self.tpm)
+
+    def validate(self):
+        """Ensure the tpm is well-formed."""
+        raise NotImplementedError
+
+    def condition_tpm(self, fixed_nodes, state):
+        """Return a TPM conditioned on the given fixed node indices, whose
+        states are fixed according to the given state-tuple.
+        """
+        raise NotImplementedError
+
+    def marginalize_out(self, node_indices):
+        """Return a TPM marginalized with respect to some nodes."""
+
+    def __repr__(self):
+        raise NotImplementedError
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __hash__(self):
+        return self._hash
+
+
+class ExplicitTPM(TPM):
+    """An explicit network TPM in multidimensional form."""
+    def __init__(self, tpm, validate=True):
+        super().__init__(tpm, validate=validate)
+        self._hash = np_hash(self.tpm)
+
+    def _build_tpm(self, tpm, validate=True):
+        """Validate the TPM passed by the user and convert to
+        multidimensional form."""
+        tpm = np.array(tpm)
+
+        if validate:
+            self._validate(
+                tpm,
+                check_independence=config.VALIDATE_CONDITIONAL_INDEPENDENCE
+            )
+
+            # Convert to multidimensional state-by-node form
+            if is_state_by_state(tpm):
+                tpm = convert.state_by_state2state_by_node(tpm)
+            else:
+                tpm = convert.to_multidimensional(tpm)
+
+        np_immutable(tpm)
+
+        return tpm
+
+    @property
+    def shape(self):
+        return self.tpm.shape
+
+    @classmethod
+    def _conditionally_independent(cls, tpm):
```

Comment:
These validation methods can be normal methods rather than class methods, and can be called after building the TPM in the constructor